File: src/main/java/com/better/suanfa/QuickSort.java
package com.better.suanfa;

import com.better.Utils;

/**
 * 快速排序
 * Created by zhaoyu on 16/9/28.
 */

public class QuickSort {
	public static void main(String[] args) {
		//int list[] = {6, 5, 2, 1, 7, 4, 5, 8, 3};
		//int list[] = {2,5,3};

		int[] list = {1, 3, 4, 5, 2, 6, 9, 7, 8, 0};

		Utils.print("排序前\t\t");
		Utils.printArray(list);

		QuickSort q = new QuickSort();
		//q.quickSort(list, 0, list.length - 1);
		q.quickSort2(list, 0, list.length - 1);

		Utils.print("排序后\t\t");
		Utils.printArray(list);
	}


	private void quickSort2(int a[], int left, int right) {
		// 前置检查
		if (left > right) {
			return;
		}


		// 记录,参数值
		int i, j, baseValue, t;
		baseValue = a[left];
		i = left;
		j = right;

		while (i < j) {
			// 从右往左,查找大于base的数 (下标 j 对应的数)
			while (a[j] >= baseValue && i < j) {
				j--;
			}

			// 从左往右,查找小于base的数 (下标 i 对应的数)
			while (a[i] <= baseValue && i < j) {
				i++;
			}

			// 【交换部分】交换2个数
			if (i < j) {
				t = a[i];
				a[i] = a[j];
				a[j] = t;
			}
		}

		// 【交换部分】基数归位
		a[left] = a[i];
		a[i] = baseValue;

		System.out.format("base = %d:\t", a[i]);
		printPart(a, left, right);

		// 分治,减小区间
		quickSort2(a, left, i - 1);        // 处理左边的
		quickSort2(a, i + 1, right);        // 处理右边的

		//Utils.println(String.format("i=%s, j=%s, base=%s, left=%s, right=%s", i, j, base, a[i], a[j]));
	}


	// 打印序列
	public void printPart(int[] list, int begin, int end) {
		for (int i = 0; i < begin; i++) {
			System.out.print("\t");
		}
		for (int i = begin; i <= end; i++) {
			System.out.print(list[i] + "\t");
		}
		System.out.println();
	}


	/* ============= sort */

	/**
	 * 快速排序
	 *
	 * @param a
	 * @param left
	 * @param right
	 */
	private void quickSort(int a[], int left, int right) {
		if (left > right) {
			return;
		}

		int i = left;
		int j = right;
		int baseValue = a[left];        // 对比基数
		int temp;

		while (i < j) {
			// 从右往左走,查找比baseValue大的数
			while (a[j] >= baseValue && i < j) {
				j--;
			}

			// 从左往右走,查找比baseValue小的数
			while (a[i] <= baseValue && i < j) {
				i++;
			}

			// 交换2个 找到的数
			if (i < j) {
				temp = a[i];
				a[i] = a[j];
				a[j] = temp;
			}
		}

		// 每次查找完毕时,交换 left 与 baseValue的值, 这时, baseValue 左边比它都小,
		// 右边比它都大
		a[left] = a[i];
		a[i] = baseValue;

		// 分治,分别快速排序
		quickSort(a, left, i-1);
		quickSort(a, i+1, right);
	}


}

File: src/main/kotlin/vip/sonar/suanfa/sort/QuickSort.kt
package vip.sonar.suanfa.sort

/**
 * @description: 快排
 * @author better
 * @date 2019-03-10 09:00
 */
fun main() {

    /**
     * 获取分区点，之前的
     */
    fun partition(a: Array<Int>, left: Int, right: Int): Int {
        val base = a[left]      // 从右开始
        var i = left
        var j = right
        while (i < j) {
            while (a[j] <= base && i < j) {   // 需使用 <=
                j--
            }
            while (a[i] >= base && i < j) {  // 需使用 >=
                i++
            }

            // if (i < j) then swap it
            if (i < j) {
                val tmp = a[i]
                a[i] = a[j]
                a[j] = tmp
            }
        }

        // 基数归位
        a[left] = a[i]
        a[i] = base
        return i
    }

    /**
     * 来自极客的，分为2部分，已处理区间与未处理区间，类似于选择排序
     * 通过游标 i 将 数组分为2部分：
     * 1. a[left, i-1] 都是小于base的
     * 2. a[i+1, right-1] 未处理的
     */
    fun partition2(a: Array<Int>, left: Int, right: Int): Int {
        val base = a[right]
        var i = left
        for (j in (left until right)) {
            if (a[j] < base) {  // 从左到右
                if (i == j) {    // 避免正序时，如 1，2，3，4 连续交换
                    ++i
                } else {
                    val tmp = a[i]
                    a[i++] = a[j]
                    a[j] = tmp
                }
            }
        }

        // 交换
        val tmp = a[i]
        a[i] = a[right]
        a[right] = tmp

        return i
    }


    /***
     * 继续编写,通过游标 i 将 数组 a 分为2部分 [left, i-1] 比 a[i] 小
     * [i+1, right-1] 未知
     */
    fun partition3(a: Array<Int>, left: Int, right: Int): Int {
        val base = a[right]
        var i = left
        for (j in (left until right)) {
            if (a[j] < base) {
                if (i == j) {  // 都小于时，避免多余的swap交换
                    i++
                } else {
                    val tmp = a[i]
                    a[i] = a[j]
                    a[j] = tmp
                    i++
                }
            }
        }

        // base 交互
        val tmp = a[i]
        a[i] = base
        a[right] = tmp
        return i
    }


    fun quickSortPart(a: Array<Int>, left: Int, right: Int) {
        if (left >= right) {
            return
        }
        // 分区
        val p = partition3(a, left, right)
        quickSortPart(a, left, p - 1)
        quickSortPart(a, p + 1, right)
    }

    fun quickSort(a: Array<Int>) {
        quickSortPart(a, 0, a.size - 1)
    }

    val array = arrayOf(1, 5, 4, 2, 0, 3, -9)
//    val array = arrayOf(0,1,2,3,4,5)
    quickSort(array)
    println(array.joinToString(" "))
}